src/core/radio/trel_packet.hpp

```diff
@@ -0,0 +1,373 @@
+     * @param[in] aPanId   A PAN Identifier.
+     *
+     */
+    void SetPanId(Mac::PanId aPanId) { mPanId = Encoding::BigEndian::HostSwap16(aPanId); }
+
+    /**
+     * This method gets the packet number field from the header.
+     *
+     * @returns The packet number field.
+     *
+     */
+    uint32_t GetPacketNumber(void) const { return Encoding::BigEndian::HostSwap32(mPacketNumber); }
+
+    /**
+     * This method sets the packet number field in the header.
+     *
+     * @param[in] aPacketNumber  The packet number.
+     *
+     */
+    void SetPacketNumber(uint32_t aPacketNumber) { mPacketNumber = Encoding::BigEndian::HostSwap32(aPacketNumber); }
+
+    /**
+     * This method gets the source MAC address field from the header.
+     *
+     * @returns The source MAC address field.
+     *
+     */
+    const Mac::ExtAddress &GetSource(void) const { return mSource; }
+
+    /**
+     * This method sets the source MAC address filed in the header.
+     *
+     * @param[in] aSource   A MAC extended address to set as source.
+     *
+     */
+    void SetSource(const Mac::ExtAddress &aSource) { mSource = aSource; }
+
+    /**
+     * This method gets the destination MAC address field from the header.
+     *
+     * This method MUST be used with a unicast of ack type packet, otherwise its behavior is undefined.
+     *
+     * @returns The destination MAC address field.
+     *
+     */
+    const Mac::ExtAddress &GetDestination(void) const { return mDestination; }
+
+    /**
+     * This method sets the destination MAC address field in the header.
+     *
+     * This method MUST be used with a unicast of ack type packet, otherwise its behavior is undefined.
+     *
+     * @param[in] aDest   A MAC extended address to set as destination.
+     *
+     */
+    void SetDestination(const Mac::ExtAddress &aDest) { mDestination = aDest; }
+
+    /**
+     * This static method gets the size (number of bytes) in header of given packet type.
+     *
+     * @param[in] aType   The packet type.
+     *
+     * @returns The fixed header size (number of bytes) for @p aType packet.
+     *
+     */
+    static uint16_t GetSize(Type aType);
+
+    /**
+     * This method returns a string representation of header.
+     *
+     * @returns An `InfoString` representation of header.
+     *
+     */
+    InfoString ToString(void) const;
+
+private:
+    enum
+    {
+        kTypeMask    = (3 << 0), // Bits 0-1 specify packet `Type`
+        kAckModeFlag = (1 << 2), // Bit 2 indicate "ack mode" (TREL ack requested or not).
+        kVersionMask = (7 << 5), // Bits 5-7 specify the version.
+        kVersion     = (0 << 5), // Current TREL header version.
+    };
+
+    // All header fields are big-endian.
+
+    uint8_t         mControl;
+    uint8_t         mChannel;
+    uint16_t        mPanId;
+    uint32_t        mPacketNumber;
+    Mac::ExtAddress mSource;
+    Mac::ExtAddress mDestination; // Present in `kTypeAck` or `kTypeUnicast` packet types.
```

Comment:
The 15.4 frame has already included the panid, source and destination address. Why do we define these fields in the  TREL radio packet?

Reply:
Note that we do need a TREL header to include a version and TREL packet number. We considered whether to include these fields (PAN Id, src and dest) in the TREL header as well or parse them from embedded 15.4 frame. 

At the end, decided to include these and have a clear and separate (well-defined) TREL header. There are multiple benefits to this. It not only helps simplify the code, it also adds flexibility to the design:
- On receive side we just need to parse the TREL header to filter TREL packet. No need to look into the encapsulated 15.4 frame.
- It adds flexibility since we can then embed any type of 15.4 frame into TREL packet, even the ones that may not have the src/dest or PAN Id info
   - For example, we can embed a 15.4 ack frame into TREL packet (note that ack frame does not contain PAN Id and/or src/dest) and ensure it is correctly filtered at the receiver at TREL link layer.
- In a 15.4 frame the src/dest adress may or may not be present and can be either short or extended. In TREL layer we would like to know the source and destination as extended addresses (since these will directly map to the (IPv6) addresses in TREL interface layer).
    - Having them as extended address in the TREL packet header would help simplify the TREL interface layer code (it does not need to look into encapsulated 15.4 frame again and be aware of neighbor table to convert from short to extended address).
   - It helps address some corner cases where it gets tricky to handle otherwise
       - We can receive a TREL packet (requiring a TREL ack) from a node which we don't yet have in our neighbor table. Having the src/dest as extended address in the TREL packet header decouples sending of TREL ack (we can prepare and send it immediately even if the neighbor is not yet added in neighbor table).
- The only benefit for not including these fields in the TREL header is to reduce the overall TREL packet size, but since TREL is run over infrastructure link with high throughput and high PSDU, optimizing few bytes does not matter much.